src/github.com/getlantern/enproxy/conn_test.go

```diff
@@ -62,18 +60,80 @@ func TestBadBuffered(t *testing.T) {
 	doTestBad(true, t)
 }
 
+func TestIdle(t *testing.T) {
+	idleTimeout := 100 * time.Millisecond
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	_, err = Dial(httpAddr, &Config{
+		DialProxy: func(addr string) (net.Conn, error) {
+			return net.Dial("tcp", proxyAddr)
+		},
+		NewRequest:  newRequest,
+		IdleTimeout: idleTimeout,
+	})
+	if assert.NoError(t, err, "Dialing should have succeeded") {
+		time.Sleep(idleTimeout * 2)
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+	}
+}
+
+// This test stimulates a connection leak as seen in
+// https://github.com/getlantern/lantern/issues/2174.
+func TestHTTPRedirect(t *testing.T) {
+	startProxy(t)
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(network, addr string) (net.Conn, error) {
+				return Dial(addr, &Config{
+					DialProxy: func(addr string) (net.Conn, error) {
+						return net.Dial("tcp", proxyAddr)
+					},
+					NewRequest: newRequest,
+				})
+			},
+			DisableKeepAlives: true,
+		},
+	}
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	resp, err := client.Head("http://www.facebook.com")
+	if assert.NoError(t, err, "Head request to facebook should have succeeded") {
+		resp.Body.Close()
+	}
+
+	assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+}
+
 func doTestPlainText(buffered bool, t *testing.T) {
 	startServers(t)
 
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
 	conn, err := prepareConn(httpAddr, buffered, false, t)
 	if err != nil {
 		t.Fatalf("Unable to prepareConn: %s", err)
 	}
-	defer conn.Close()
+	defer func() {
+		err := conn.Close()
+		assert.Nil(t, err, "Closing conn should succeed")
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
```

Comment:
So far I've found that the test doest not fail when we have these matching file descriptors:

```
startingLines: enproxy.t 2867 reventlov    5u    IPv4 0x6510f8cd1e78ed25       0t0     TCP localhost:59779 (LISTEN)
enproxy.t 2867 reventlov    8u    IPv4 0x6510f8cd1fb132b5       0t0     TCP localhost:59790 (LISTEN)
enproxy.t 2867 reventlov    9u    IPv4 0x6510f8cd1f28d795       0t0     TCP localhost:59792 (LISTEN)
```

But, sometimes a CLOSE_WAIT appears:

```
enproxy.t 2914 reventlov    5u    IPv4 0x6510f8cd1ce7cb85       0t0     TCP localhost:60430 (LISTEN)
enproxy.t 2914 reventlov    8u    IPv4 0x6510f8cd1efbaec5       0t0     TCP localhost:60456 (LISTEN)
enproxy.t 2914 reventlov    9u    IPv4 0x6510f8cd1f2f1795       0t0     TCP localhost:60458 (LISTEN)
enproxy.t 2914 reventlov   10u    IPv4 0x6510f8cd1fae65f5       0t0     TCP localhost:60458->localhost:60459 (CLOSE_WAIT)
```

and the test fails.

The CLOSE_WAIT fd is a temporary status, in fact, if I add a `time.Sleep(time.Second * 1)` after `startServers()` on the `doTestPlainText()` function the test does not fail anymore so my guess is that this is something that can be fixed within `startServers()`, which I'm now reviewing.